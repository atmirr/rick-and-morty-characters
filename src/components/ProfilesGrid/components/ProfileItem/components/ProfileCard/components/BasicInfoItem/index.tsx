import React from 'react';
import styled from 'styled-components';

export const statusColors: { [key: string]: string } = {
  alive: '#55cc44',
  dead: '#d53d2e',
  unknown: '#9e9e9e',
};

const Container = styled.div`
  display: flex;
  align-items: center;
  flex-wrap: wrap;
  font-size: 16px;
  line-height: 1.4em;
  padding-bottom: 2px;
  color: #797979;
`;

const Divider = styled.span`
  padding: 0 3px;
`;

const Wrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
`;

const Status = styled.span`
  text-transform: capitalize;
`;

const StatusIcon = styled.span`
  height: 7px;
  width: 7px;
  margin-right: 4px;
  background-color: ${({ status }: { status: string }) => statusColors[status]};
  border-radius: 50%;
`;

const Species = styled.span`
  text-transform: capitalize;
`;

const Gender = styled.span`
  padding-left: 3px;
  font-size: 13px;
`;

type PropTypes = {
  gender: string;
  status: string;
  species: string;
};

function BasicInfoItem({
  species,
  status,
  gender,
}: PropTypes): React.ReactElement {
  return (
    <Container>
      <Wrapper>
        <StatusIcon data-testid="status-icon" status={status?.toLowerCase()} />
        <Status>{status}</Status>
      </Wrapper>
      <Divider>-</Divider>
      <Wrapper>
        <Species>{species}</Species>
        <Gender>({gender})</Gender>
      </Wrapper>
    </Container>
  );
}

export default BasicInfoItem;
