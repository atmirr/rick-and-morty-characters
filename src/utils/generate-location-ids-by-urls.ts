export const generateLocationIdsByUrls = (urls: string[]): string[] =>
  urls.reduce((result: string[], url: string) => {
    const splittedUrl = url.split('/');
    const id = splittedUrl[splittedUrl.length - 1];
    return id ? [...result, id] : [...result];
  }, []);
