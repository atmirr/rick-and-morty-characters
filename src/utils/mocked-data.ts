export const characters = [
  {
    id: 1,
    name: 'Rick Sanchez',
    species: 'Human',
    gender: 'Male',
    created: '2017-11-04T18:48:46.250Z',
    episode: [
      'https://rickandmortyapi.com/api/episode/1',
      'https://rickandmortyapi.com/api/episode/2',
    ],
    image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
    location: {
      name: 'Earth (Replacement Dimension)',
      url: 'https://rickandmortyapi.com/api/location/20',
    },
    origin: {
      name: 'Earth (C-137)',
      url: 'https://rickandmortyapi.com/api/location/1',
    },
    status: 'Alive',
    type: '',
    url: 'https://rickandmortyapi.com/api/character/1',
  },
  {
    id: 2,
    name: 'Morty Smith',
    species: 'Human',
    gender: 'Male',
    created: '2017-11-04T18:50:21.651Z',
    episode: [
      'https://rickandmortyapi.com/api/episode/1',
      'https://rickandmortyapi.com/api/episode/2',
      'https://rickandmortyapi.com/api/episode/3',
    ],
    image: 'https://rickandmortyapi.com/api/character/avatar/2.jpeg',
    location: {
      name: 'Earth (Replacement Dimension)',
      url: 'https://rickandmortyapi.com/api/location/20',
    },
    origin: {
      name: 'Earth (C-137)',
      url: 'https://rickandmortyapi.com/api/location/1',
    },
    status: 'Alive',
    type: '',
    url: 'https://rickandmortyapi.com/api/character/2',
  },
];

export const episodes = [
  'Rixty Minutes',
  'Something Ricked This Way Comes',
  'Close Rick-counters of the Rick Kind',
  'Ricksy Business',
  'A Rickle in Time',
];

export const locations = [
  {
    name: 'Citadel of Ricks',
    type: 'Space station',
    dimension: 'unknown',
  },
  {
    name: 'Earth (Replacement Dimension)',
    type: 'Planet',
    dimension: 'Replacement Dimension',
  },
];
