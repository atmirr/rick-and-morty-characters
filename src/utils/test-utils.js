import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

const mockedInitialState = {
  characters: { items: [], nextPage: null, error: null, isLoading: false },
  episodes: { items: [], nextPage: null, error: null, isLoading: false },
  locations: { items: [], error: null, isLoading: false },
};

function render(
  ui,
  {
    initialState = mockedInitialState,
    store = configureStore()(initialState),
    ...renderOptions
  } = {},
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from '@testing-library/react';
// override render method
export { render };
