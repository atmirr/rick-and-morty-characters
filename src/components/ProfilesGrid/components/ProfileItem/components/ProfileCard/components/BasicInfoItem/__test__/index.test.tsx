import React from 'react';
import 'jest-styled-components';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';
import { screen, render } from 'src/utils/test-utils';
import BasicInfoItem from '../index';

const props = {
  species: 'Human',
  gender: 'Male',
  status: 'Alive',
};

describe('<BasicInfoItem />', () => {
  it('should renders the component with the given props', () => {
    const { container } = render(<BasicInfoItem {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should display species this props has been passed', () => {
    render(<BasicInfoItem species={props.species} />);
    expect(screen.queryAllByText(props.species).length).toBe(1);
  });

  it('should display gender this props has been passed', () => {
    render(<BasicInfoItem gender={props.gender} />);
    expect(screen.queryAllByText(`(${props.gender})`).length).toBe(1);
  });

  it('should display status when this props has been passed', () => {
    render(<BasicInfoItem status={props.status} />);
    expect(screen.queryAllByText(props.status).length).toBe(1);
  });
});
