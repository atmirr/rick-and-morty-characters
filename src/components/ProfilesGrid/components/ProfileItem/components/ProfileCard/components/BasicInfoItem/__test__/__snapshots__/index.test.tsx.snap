// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BasicInfoItem /> should renders the component with the given props 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  font-size: 16px;
  line-height: 1.4em;
  padding-bottom: 2px;
  color: #797979;
}

.c4 {
  padding: 0 3px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 {
  text-transform: capitalize;
}

.c2 {
  height: 7px;
  width: 7px;
  margin-right: 4px;
  background-color: #55cc44;
  border-radius: 50%;
}

.c5 {
  text-transform: capitalize;
}

.c6 {
  padding-left: 3px;
  font-size: 13px;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <span
      class="c2"
      data-testid="status-icon"
    />
    <span
      class="c3"
    >
      Alive
    </span>
  </div>
  <span
    class="c4"
  >
    -
  </span>
  <div
    class="c1"
  >
    <span
      class="c5"
    >
      Human
    </span>
    <span
      class="c6"
    >
      (
      Male
      )
    </span>
  </div>
</div>
`;
